File: Контест 5/J.py
class Node:
    def __init__(self, key):
      self.data = key
      self.left = None
      self.right = None
      self.count = 1

class BinarySearchTree:

    def __init__(self):
        self.root = None

    def insert(self, key):
      if self.root is None:
        self.root = Node(key)
      else:
        curr_node = self.root
        while True:
          if key < curr_node.data:
            if curr_node.left is None:
              curr_node.left = Node(key)
              break
            curr_node = curr_node.left
          elif key > curr_node.data:
            if curr_node.right is None:
              curr_node.right = Node(key)
              break
            curr_node = curr_node.right
          else:
            curr_node.count += 1
            break

    def order_up(self, node):
      result = []
      stack = []
      current = self.root
      while stack or current:
        while current:
          stack.append(current)
          current = current.left
        current = stack.pop()
        item = [current.data] + [current.count]
        result.append(item)
        current = current.right
      return result
        
    def search(self, key):
      return self._search(self.root, key)

    def _search(self, node, key):
      if node is None:
          return False
      if key == node.data:
          return True
      if key < node.data:
          return self._search(node.left, key)
      return self._search(node.right, key)

numbers = list(map(int, input().split()))
bst = BinarySearchTree()
for el in numbers:
  if el == 0:
    break
  else:
    bst.insert(el)

for el in bst.order_up(bst.root):
  print(*el)

File: Контест 5/J.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

struct Node
{
  int data;
  Node *left;
  Node *right;
  int count;

  Node(int key) : data(key), left(nullptr), right(nullptr), count(1) {}
};

class BinarySearchTree
{
public:
  BinarySearchTree() : root(nullptr) {}

  void insert(int key)
  {
    if (root == nullptr)
    {
      root = new Node(key);
      return;
    }

    Node *curr_node = root;
    while (true)
    {
      if (key < curr_node->data)
      {
        if (curr_node->left == nullptr)
        {
          curr_node->left = new Node(key);
          break;
        }
        curr_node = curr_node->left;
      }
      else if (key > curr_node->data)
      {
        if (curr_node->right == nullptr)
        {
          curr_node->right = new Node(key);
          break;
        }
        curr_node = curr_node->right;
      }
      else
      {
        curr_node->count++;
        break;
      }
    }
  }

  void order_up_recursive(Node *node, vector<pair<int, int>> &result)
  {
    if (node != nullptr)
    {
      order_up_recursive(node->left, result);
      result.push_back(make_pair(node->data, node->count));
      order_up_recursive(node->right, result);
    }
  }

  vector<pair<int, int>> order_up()
  {
    vector<pair<int, int>> result;
    order_up_recursive(root, result);
    return result;
  }

  Node *getRoot()
  { 
    return root;
  }

  ~BinarySearchTree()
  {
    queue<Node *> nodes;
    nodes.push(root);
    while (!nodes.empty())
    {
      Node *current = nodes.front();
      nodes.pop();
      if (current->left)
        nodes.push(current->left);
      if (current->right)
        nodes.push(current->right);
      delete current;
    }
  }

private:
  Node *root;
};

int main()
{
  vector<int> numbers;
  int el;
  cin >> el;
  while (el != 0)
  {
    numbers.push_back(el);
    cin >> el;
  }

  BinarySearchTree bst;
  for (int num : numbers)
  {
    bst.insert(num);
  }

  vector<pair<int, int>> result = bst.order_up();
  for (auto item : result)
  {
    cout << item.first << " " << item.second << endl;
  }
  return 0;
}